### PR TITLE
allow biography length configuration

### LINKED ISSUE #20: Configurable bio character counts
Many glitch.social forks seem to have this but it's not documented anywhere. Keeping this open until we have things translated and an API is set up.

Example on the cybre.space fork:
* https://cybre.tech/cybrespace/mastodon/commit/a2aea63ae9306d4f29713a6e31bcf263dd775afc

### PR DESCRIPTION
**Pull request by ghost**
Closes #20 and possibly #42 

This adds an administrative setting that allows for arbitrary biography length.  It is based on cybrespace's changeset

screenshot 2019-02-27 19 48 52

### CODE REVIEW

File: config/locales/en.yml [hunk 1]
```diff
@@ -392,8 +392,11 @@ en:
       mascot:
         desc_html: Displayed on multiple pages. At least 293×205px recommended. When not set, falls back to default mascot
         title: Mascot image
+      max_bio_chars:
+        desc_html: Set maximum biography length
+        title: Max Biography Size
       max_toot_chars:
-        desc_html: 'Set maximum toot length in characters (default: 500)'
+        desc_html: Set maximum toot length
```
Comment (clarfonthey) [inline config/locales/en.yml]: Why is this reverted back?

File: config/locales/en.yml [hunk 2]
```diff
@@ -392,8 +392,11 @@ en:
       mascot:
         desc_html: Displayed on multiple pages. At least 293×205px recommended. When not set, falls back to default mascot
         title: Mascot image
+      max_bio_chars:
+        desc_html: Set maximum biography length
```
Comment (clarfonthey) [inline config/locales/en.yml]: Should use "in characters" similar to max toot length.

I'd probably rename "biography" to "user bio" here although it seems that biography is preferred over bio in other strings. Maybe "user biography"?

File: config/settings.yml [hunk 3]
```diff
@@ -10,6 +10,7 @@ defaults: &defaults
   site_contact_username: ''
   site_contact_email: ''
   open_registrations: true
+  max_bio_chars: 160
```
Comment (clarfonthey) [inline config/settings.yml]: If we're going to add this we might as well default to 500 like others have suggested.

File: app/models/account.rb [hunk 4]
```diff
@@ -75,7 +75,7 @@ class Account < ApplicationRecord
   validates_with UniqueUsernameValidator, if: -> { local? && will_save_change_to_username? }
   validates_with UnreservedUsernameValidator, if: -> { local? && will_save_change_to_username? }
   validates :display_name, length: { maximum: 30 }, if: -> { local? && will_save_change_to_display_name? }
-  validates :note, note_length: { maximum: 160 }, if: -> { local? && will_save_change_to_note? }
+  validates :note, note_length: { maximum: Setting.max_bio_chars.to_i }, if: -> { local? && will_save_change_to_note? }
```
Comment (clarfonthey) [inline app/models/account.rb:78]: Is there a reason why you added the `to_i` here?

File: app/views/settings/profiles/show.html.haml [hunk 5]
```diff
@@ -7,7 +7,7 @@
   .fields-row
     .fields-row__column.fields-group.fields-row__column-6
       = f.input :display_name, wrapper: :with_label, input_html: { maxlength: 30 }, hint: false
-      = f.input :note, wrapper: :with_label, input_html: { maxlength: 160 }, hint: false
+      = f.input :note, wrapper: :with_label, input_html: { maxlength: Setting.max_bio_chars.to_i }, hint: false
```
Comment (clarfonthey) [inline app/views/settings/profiles/show.html.haml:10]: Also here.

File: app/javascript/packs/public.js [hunk 6]
```diff
@@ -116,6 +123,16 @@ function main() {
       document.head.appendChild(scrollbarWidthStyle);
       scrollbarWidthStyle.sheet.insertRule(`body.with-modals--active { margin-right: ${scrollbarWidth}px; }`, 0);
     }
+    [].forEach.call(document.querySelectorAll('[data-component="Card"]'), (content) => {
```
Comment (clarfonthey) [inline app/javascript/packs/public.js:126]: I'm not 100% familiar with react; what exactly is this `forEach` call doing?

File: app/javascript/packs/public.js [hunk 7]
```diff
@@ -214,6 +231,22 @@ function main() {
...
+    const noteCounter = document.querySelector('.note-counter');
+    const bioTextArea = document.querySelector('#account_note');
+
+    if (noteCounter) {
+      noteCounter.textContent = maxBioChars - length(bioTextArea.value);
+    }
+
+    if (bioTextArea) {
+      bioTextArea.style.height = 'auto';
+      bioTextArea.style.height = (bioTextArea.scrollHeight+3) + 'px';
```
Comment (clarfonthey) [inline app/javascript/packs/public.js:270]: Is there any reason why there's a `3` here?

File: app/javascript/packs/public.js [hunk 8]
```diff
@@ -214,6 +231,22 @@ function main() {
       console.error(err);
     }
   });
+
+  delegate(document, '#account_note', 'input', sizeBioText);
+
+  function sizeBioText() {
```
Comment (clarfonthey) [inline app/javascript/packs/public.js:260]: Does the bio not already have a length counter in mainline? I feel like it might make sense to separate this out into another commit if that's the case.

Although Gargron isn't going to merge a configurable length, I feel like a length counter would be merged into mainline. Perhaps it should at least be separated out into a separate commit than the configuration itself.

**Comment by clarfonthey** [comment]
Added my thoughts on the review. Would you be able to provide a screenshot of the length counter in the editor as well, as this modifies that too?

**Comment by ghost** [comment]
@clarfon The biography section does not display a counter today.  The javascript changes you see around bio length have to do with setting the height of the edit box based on max length.  I have these changes live on my [instance](https://banana.dog) (actually I usually roll out all changes to my instance before I PR them here)

**Comment by igalic** [comment]
i'll be honest, i totally expected this to be database bound.

**Review by renatolond — DISMISSED** [review]

File: config/settings.yml (hunk 3 above)
Comment (1011X) [inline config/settings.yml]: If this is changed to 500, #42 can be closed.
```suggestion
  max_bio_chars: 500
```

**Review by 1011X — CHANGES_REQUESTED** [review]

File: config/settings.yml (hunk 3 above)
Comment (ghost) [reply config/settings.yml]: Hold on I can update it...

**Review by 1011X — DISMISSED** [review]

**Comment by 1011X** [comment]
Okay, some tests are failing because they expect an error for creating or submitting a long note. The reason for that is there are hard-coded values/expectations in the tests that don't align with the config values provided in other files. The tests are:

+ [/spec/models/account_spec.rb:602](https://github.com/florence-social/mastodon-fork/blob/master/spec/models/account_spec.rb#L599-L603)
+ [/spec/controllers/api/v1/accounts/credentials_controller_spec.rb:68](https://github.com/florence-social/mastodon-fork/blob/master/spec/controllers/api/v1/accounts/credentials_controller_spec.rb#L62-L70)

Possible solutions in increasing difficulty:
+ Remove the tests
+ Modify tests so they pass according to the values we expect (still hard-coded)
+ Modify tests so they pass according to the values specified in the configuration

<details><summary>Relevent test failure output</summary>

```
  1) Account validations when is local is invalid if the note is longer than 160 characters
     Failure/Error: expect(account).to model_have_error_on_field(:note)
       expect record.errors([]) to include note
     # ./spec/models/account_spec.rb:602:in `block (4 levels) in <top (required)>'

  1) Api::V1::Accounts::CredentialsController with an oauth token PATCH #update with invalid data returns http unprocessable entity
     Failure/Error: expect(response).to have_http_status(:unprocessable_entity)
       expected the response to have status code :unprocessable_entity (422) but it was :ok (200)
     # ./spec/controllers/api/v1/accounts/credentials_controller_spec.rb:68:in `block (5 levels) in <top (required)>'
```
</details>

**Comment by renatolond** [comment]
If the value is configurable, then the tests can configure a value and test that notes above that value are not accepted. There's examples of settings being reconfigured for tests in `spec/controllers/auth/registrations_controller_spec.rb` and `spec/controllers/application_controller_spec.rb`, in slightly different ways.

**Comment by ghost** [comment]
I am updating the tests now!

**Review by 1011X — DISMISSED** [review]

**Review by jhaye — CHANGES_REQUESTED** [review]
Everything looks fine - would you mind rebasing?

**Comment by ghost** [comment]
@jhaye squashed and rebased as requested

**Review by jhaye — DISMISSED** [review]

**Comment by clarfonthey** [comment]
@usbsnowcrash can you update the locale strings for the bio char count too? Thanks

File: app/helpers/application_helper.rb [hunk 9]
```diff
@@ -39,6 +39,14 @@ def available_sign_up_path
     end
   end
 
+  def max_bio_chars
+    Setting.max_bio_chars
+  end
+
+  def max_bio_chars
+    Setting.max_bio_chars
+  end
```
Comment (1011X) [inline app/helpers/application_helper.rb]: Is there supposed to be 2 copies of the same function?

Aside from that, everything here looks good.

Comment (ghost) [reply app/helpers/application_helper.rb]: No, looks like a bad merge.  I'll remove the dupe

**Comment by ghost** [comment]
> @usbsnowcrash can you update the locale strings for the bio char count too? Thanks

Sure thing

**Review by clarfonthey — APPROVED** [review]

**Comment by clarfonthey** [comment]
+1 from me, whoever does the next +1 can merge.

File: app/javascript/packs/public.js (hunk 6 above)
Comment (1011X) [reply app/javascript/packs/public.js:126]: [`.forEach()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach) is native JavaScript.

Comment (clarfonthey) [reply app/javascript/packs/public.js:126]: I know that, I meant the purpose of this loop. :sweat_smile:

File: app/javascript/packs/public.js (hunk 8 above)
Comment (ghost) [reply app/javascript/packs/public.js:260]: Bio does not have a size counter today and this function just resizes the box dynamically to ensure it is y’all enough to accommodate the bio area.

File: app/models/account.rb (hunk 4 above)
Comment (ghost) [reply app/models/account.rb:78]: When the setting has never been set before in the database and it comes from the default yaml it will emerge from the ruby side as a string and this just coerces it so the JavaScript is an explicit int just like it was when it was hard coded to 160.

File: app/javascript/packs/public.js (hunk 6 above)
Comment (1011X) [reply app/javascript/packs/public.js:126]: Oh. That I don't know.

File: app/javascript/packs/public.js (hunk 7 above)
Comment (ghost) [reply app/javascript/packs/public.js:270]: I don’t remember why.  I believe this comes directly from the cybrespace changesets but I don’t know why it’s 3px.

File: app/javascript/packs/public.js (hunk 6 above)
Comment (ghost) [reply app/javascript/packs/public.js:126]: I believe this is a trick to ensure you can use the foreach even in cases where a single value comes back instead of an array...

**Review by 1011X — APPROVED** [review]
